x/config/doc.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package config implements a generic configuration system that may be used
+// to build YARPC Dispatchers from configurations specified in different
+// markup formats.
+//
+// Usage
+//
+// To build a Dispatcher, set up a Loader and inform it about the different
+// transports that it needs to support. Use Load or LoadYAML to load a Builder
+// from the given configuration.
+//
+// 	loader := config.NewLoader()
+// 	http.RegisterTransport(loader)
+// 	redis.RegisterTransport(loader)
```

Comment:
hrrrrrrrrrrm... so having the RegisterTransport functions on HTTP and Redis have side effects on the loader makes it harder to understand when/why the loader's state changes.  We should probably invert the logic so the loader gets configuration from the transports
```
loader.RegisterTransport(http.TransportLoader())
loader.RegisterTransport(redis.TransportLoader())
```
or pass in the transports as an option to the loader
```
loader := config.NewLoader(
  config.TransportLoader(http.TransportLoader()),
  config.TransportLoader(redis.TransportLoader()),
)
```